src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs

```diff
@@ -167,6 +168,22 @@ public async Task SslStream_NetworkStream_Renegotiation_Succeeds()
             }
         }
 
+        [Fact]
+        public async Task SslStream_NestedAuth_Throws()
+        {
+            VirtualNetwork network = new VirtualNetwork();
+
+            using (var clientStream = new VirtualNetworkStream(network, isServer: false))
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var ssl = new SslStream(clientStream))
+            {
+                // Start handshake.
+                Task task = ssl.AuthenticateAsClientAsync("foo.com", null, SslProtocols.Tls12, false);
+                // Do it again without waiting for previous one to finish.
```

Comment:
In most cases, the previous task will not have finished, so the test itself will probably succeed.

But this test is an example of hoping that race conditions favor us and leads us towards unstable tests.

This test raises a few questions in my mind.

If the task finished quickly but failed, then the subsequent .AuthenticateAsClientAsync will be invoked and probably not throw an InvalidOperationException. So, the test itself will fail.

If the task finished quickly as successful then will the subsequent .AuthenticateAsClientAsync fail with InvalidOperationException? I don't recall if we allow SslStream to re-do .AuthenticateAsClient*. I'm guessing we don't. So, will it throw InvalidOperationException?

Reply:
This really should not happen. Since there is no server, first request will write ClientHello to memory buffer and it will wait for response which will never come.